service/history/shard/context_test.go

```diff
@@ -711,3 +711,46 @@ func (s *contextSuite) TestShardStopReasonCloseShard() {
 	s.False(s.mockShard.IsValid())
 	s.False(s.mockShard.stoppedForOwnershipLost())
 }
+
+func (s *contextSuite) TestUpdateShardInfo_PersistsAfterInterval() {
+	s.mockShard.state = contextStateAcquired
+	s.mockShardManager.EXPECT().UpdateShard(gomock.Any(), gomock.Any()).
+		Return(nil).Times(2)
+
+	var timesCalled int
+	callback := func() {
+		timesCalled++
+	}
+	err := s.mockShard.updateShardInfo(callback)
+	s.NoError(err)
+
+	// No time has passed: shouldn't update
```

Comment:
Huh, as I was reading this comment, I didn't expect `timesCalled` to be 3 at the end since the next run simulates passing time. I think I don't understand the purpose of this comment then.

Reply:
This means we shouldn't update the database. The callback should be called _every_ time `updateShardInfo` is, but the changes should only be persisted to the database twice during the first and final calls.

Reply:
All of these test the persistence logic; I'm asserting that we're still calling the in-memory update logic _even when_ we aren't persisting to the DB on that call to updateShardInfo

Reply:
Oh, I get it. So that's in regards to `Times(2)`. And the callback is always invoked, that's `3`.

I think separate sub-test cases could help make that clearer. Or having an individual `s.mockShardManager.EXPECT()` right above each call that updates (I actually don't know the limitations of gomock, not sure if that works).

Reply:
It looks like that does work, so I've both added a test to specifically assert that the callback is always called and have updated the other test to specify whether the DB should be hit before each invocation

Reply:
Neat! ❤️